### DISCUSSION TITLE
Question: HTTP Request Log

### DISCUSSION THREAD

**Discussion by a1730** [activej/activej#175] [Q&A]
Hello,
Does ActiveJ have facilities for HTTP request logging aka Access log? 
We have looked but not found anything in the sources so far.
We will appreciate any pointers.
Thank you.

**Comment by eduard-vasinskyi** [comment]
Hello, @a1730
You can write a root [AsyncServlet](https://github.com/activej/activej/blob/master/core-http/src/main/java/io/activej/http/AsyncServlet.java) that wraps around the actual servlet and logs all of the incoming HTTP requests as well as server HTTP responses. Pass this wrapper directly to the `AsyncHttpServer#create` method.

**Comment by a1730** [comment]
Hello @eduard-vasinskyi ,
Excellent.
We were hoping that we were missing something that's already being used by others:)
We will look at implementing something over the weekend along the lines that you suggested in #67.
Thank you.

**Comment by a1730** [comment]
Hello @eduard-vasinskyi,

We fund that we could use [AsyncServletDecorator](https://github.com/activej/activej/blob/master/core-http/src/main/java/io/activej/http/AsyncServletDecorator.java) using logged with our implementation of [LoggableServlet.](https://github.com/activej/activej/blob/master/core-http/src/main/java/io/activej/http/LoggableServlet.java) This has been working well for us. We could not use LoggableServlet as provided because we wanted the flexibility to change our log level.

Can you please comment on our current approach?

**Comment by a1730** [comment]
Hello @eduard-vasinskyi,

I forgot to mention that we sometimes get a NULL response (HTTPResponse) reference in our logger even when the wrapped servlet is successful i.e 200-OK.
Consider the following
```  @Provides
  AsyncServlet servlet(Executor executor) {
    AsyncServletDecorator combineDecorators =
      combineDecorators(List.of(loadBody(), logged(new CustomRequestLog()::log)));
    AsyncServlet router = RoutingServlet.create().map(GET, "/*", request -> {
      return HttpResponse.ok200().withPlainText("Hi there");
    }).then(catchRuntimeExceptions()).then(
      mapException(e -> HttpResponse.ofCode(404).withPlainText("Error: " + e)));
    return combineDecorators.serve(router);
  }
```
Why would response handle in logged be null? We figure we are missing something about the lifecycle of those servlets. We currently test for null in our logger.  However, we have not successfully reproduced the [common log format](https://en.wikipedia.org/wiki/Common_Log_Format) because we do not have a way to calculate **the size of the object returned to the client, measured in bytes** using the HTTPResponse object.

What are we missing?
Thank you and your team for ActiveJ.

**Comment by eduard-vasinskyi** [comment]
Hello, @a1730 
If you followed code from [LoggableServlet](https://github.com/activej/activej/blob/master/core-http/src/main/java/io/activej/http/LoggableServlet.java) when developing your own logging decorator, the request may be `null` in the case when the actual servlet returns a response asynchronously. As `LoggableServlet` additionally logs the request once it arrives even if the response has not been received yet. What you can do is:
1. Read and store the data required for logs from the request once it arrives
2. Pass the request to the actual servlet to serve
3. After it finishes - read the data required for logs from the response (which is not `null`)
4. Log the combined data obtained from the request and the response
 
As for the size of the object returned to the client, you can do this:
```
if (httpResponse.isBodyLoaded()) {
  responseBodySize = httpResponse.getBody().readRemaining();
}
```

**Comment by a1730** [comment]
FWIW we used [CustomRequestLog](https://github.com/eclipse/jetty.project/blob/3d70c8b4abaa689d5fe4a12c3aeaf2702dd4af68/jetty-server/src/main/java/org/eclipse/jetty/server/CustomRequestLog.java) from the Jetty project as an extensible base for our implementation. It has served us well.
